Fix/refreshing tokens - expiry in inactive tabs - core refactor

uvdsl opened the pull request:
This PR is triggered by #13. Thank you, @angelo-v and @jg10-mastodon-social, for your feedback :)

Basically, I decoupled the core functionality from the specific implementation using the IndexedDB (and Web Worker).
So, as mentioned in the [wiki](https://github.com/uvdsl/solid-oidc-client-browser/wiki/API-Reference#library-versions-web-vs-core), you can also implement the token refresh lifecycle yourself. Just offering more flexibility while trying to provide a nice out-of-the-box experience...

The out-of-the-box `web` version uses a Web Worker to maintain a single session state that is initialised by the first tab,  distributed across tabs, and automatically refreshed.

This PR should (TM) address the stale session issue that occurs after hibernation.

uvdsl committed:
initial refactor of structure re session database

uvdsl committed:
build tool

uvdsl committed:
add tests

uvdsl committed:
update bundle

uvdsl committed:
audit fix

uvdsl committed:
added some test, omitted 'pita' tests

uvdsl committed:
minor polish

uvdsl committed:
fix tests

uvdsl committed:
WebWorker: do not stop the timers when no ports open, just happily do your thing

uvdsl committed:
trying to fix inactive tabs

uvdsl committed:
adding onStateChange callback because vue's reactivity systems does not track state updates triggered from the web worker

uvdsl committed:
check token expiration to not provide stale tokens

angelo-v commented on coverage/lcov-report/core/AuthorizationCodeGrant.ts.html:
```diff
@@ -0,0 +1,1003 @@
+
```
Did you check in the coverage reports deliberately? I usually add them to gitignore

uvdsl committed:
minor error polish

angelo-v commented on src/web/RefreshWorker.ts:
```diff
@@ -0,0 +1,129 @@
+let tokenDetails: TokenDetails | undefined = undefined;
+let exp: number | undefined;
+
+self.onconnect = (event: any) => {
```
Regarding testability: You could extract the onconnect function to its own file to unit test it. The timeouts can be tested using jest.useFakeTimers()

uvdsl committed:
refactored session classes

- moved token expiration check on authFetch into core
- moved onSessionStateChange into core
- test behaviour manually in test app

TODO unit testing ...

uvdsl committed:
added tests for modified SessionCore

uvdsl committed:
add tests

uvdsl committed:
remove coverage

uvdsl committed:
update gitignore

uvdsl committed:
refreshing tokens promise check

uvdsl committed:
support proper tab sync on login and logout

uvdsl committed:
testing and fixing

uvdsl committed:
updated readme
